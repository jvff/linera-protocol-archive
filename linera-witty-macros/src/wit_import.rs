// Copyright (c) Zefchain Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

//! Generation of code to import functions from a Wasm guest module.

use proc_macro2::{Span, TokenStream};
use proc_macro_error::abort;
use quote::{quote, quote_spanned, ToTokens};
use syn::{spanned::Spanned, FnArg, Ident, ReturnType, TraitItem, TraitItemMethod};

/// Pieces of information extracted from a function's definition.
struct FunctionInformation<'input> {
    function: &'input TraitItemMethod,
    parameter_definitions: TokenStream,
    parameter_bindings: TokenStream,
    return_type: TokenStream,
    interface: TokenStream,
    instance_constraint: TokenStream,
}

impl<'input> FunctionInformation<'input> {
    /// Extracts the necessary information from the `function` and stores it in a new
    /// [`FunctionInformation`] instance.
    pub fn new(function: &'input TraitItemMethod) -> Self {
        let (parameter_definitions, parameter_bindings, parameter_types) =
            Self::parse_parameters(function.sig.inputs.iter());

        let return_type = match &function.sig.output {
            ReturnType::Default => quote_spanned! { function.sig.output.span() => () },
            ReturnType::Type(_, return_type) => return_type.to_token_stream(),
        };

        let interface = quote_spanned! { function.sig.span() =>
            <(linera_witty::HList![#parameter_types], #return_type)
                as linera_witty::ImportedFunctionInterface>
        };

        let instance_constraint = quote_spanned! { function.sig.span() =>
            linera_witty::InstanceWithFunction<
                #interface::GuestParameters,
                #interface::GuestResults,
            >
        };

        FunctionInformation {
            function,
            parameter_definitions,
            parameter_bindings,
            return_type,
            interface,
            instance_constraint,
        }
    }

    /// Parses a function's parameters and returns the pieces constructed from the parameters.
    ///
    /// Returns the parameter definitions (the name and type pairs), the parameter bindings (the
    /// names) and the parameter types.
    fn parse_parameters(
        function_inputs: impl Iterator<Item = &'input FnArg>,
    ) -> (TokenStream, TokenStream, TokenStream) {
        let parameters = function_inputs.map(|input| match input {
            FnArg::Typed(parameter) => parameter,
            FnArg::Receiver(receiver) => abort!(
                receiver.self_token,
                "Imported interfaces can not have `self` parameters"
            ),
        });

        let mut parameter_definitions = quote! {};
        let mut parameter_bindings = quote! {};
        let mut parameter_types = quote! {};

        for parameter in parameters {
            let parameter_binding = &parameter.pat;
            let parameter_type = &parameter.ty;

            parameter_definitions.extend(quote! { #parameter, });
            parameter_bindings.extend(quote! { #parameter_binding, });
            parameter_types.extend(quote! { #parameter_type, });
        }

        (parameter_definitions, parameter_bindings, parameter_types)
    }

    /// Returns the name of the function.
    pub fn name(&self) -> &Ident {
        &self.function.sig.ident
    }

    /// Returns the code span of the function.
    pub fn span(&self) -> Span {
        self.function.span()
    }
}

impl<'input> From<&'input TraitItem> for FunctionInformation<'input> {
    fn from(item: &'input TraitItem) -> Self {
        match item {
            TraitItem::Method(function) => FunctionInformation::new(function),
            TraitItem::Const(const_item) => abort!(
                const_item.ident,
                "Const items are not supported in imported traits"
            ),
            TraitItem::Type(type_item) => abort!(
                type_item.ident,
                "Type items are not supported in imported traits"
            ),
            TraitItem::Macro(macro_item) => abort!(
                macro_item.mac.path,
                "Macro items are not supported in imported traits"
            ),
            _ => abort!(item, "Only function items are supported in imported traits"),
        }
    }
}
