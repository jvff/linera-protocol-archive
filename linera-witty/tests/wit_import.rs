// Copyright (c) Zefchain Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

//! Tests for the `wit_import` attribute macro.

#![allow(clippy::bool_assert_comparison)]

#[path = "common/test_instance.rs"]
mod test_instance;

use self::test_instance::{MockInstanceFactory, TestInstanceFactory};
use linera_witty::{Instance, Runtime, RuntimeMemory};
use linera_witty_macros::wit_import;
use test_case::test_case;

/// An interface to import a single function without parameters or return values.
#[wit_import(package = "witty-macros:test-modules")]
trait SimpleFunction {
    fn simple();
}

/// Test importing a simple function without parameters or return values.
#[test_case(MockInstanceFactory::default(); "with a mock instance")]
fn simple_function<InstanceFactory>(mut factory: InstanceFactory)
where
    InstanceFactory: TestInstanceFactory,
    InstanceFactory::Instance: InstanceForSimpleFunction,
    <<InstanceFactory::Instance as Instance>::Runtime as Runtime>::Memory:
        RuntimeMemory<InstanceFactory::Instance>,
{
    let instance = factory.load_test_module("simple-function");

    SimpleFunction::new(instance)
        .simple()
        .expect("Failed to call guest's `simple` function");
}

/// An interface to import functions with return values.
#[wit_import(package = "witty-macros:test-modules")]
trait Getters {
    fn get_true() -> bool;
    fn get_false() -> bool;
    fn get_s8() -> i8;
    fn get_u8() -> u8;
    fn get_s16() -> i16;
    fn get_u16() -> u16;
    fn get_s32() -> i32;
    fn get_u32() -> u32;
    fn get_s64() -> i64;
    fn get_u64() -> u64;
    fn get_float32() -> f32;
    fn get_float64() -> f64;
}

/// Test importing functions with return values.
#[test_case(MockInstanceFactory::default(); "with a mock instance")]
fn getters<InstanceFactory>(mut factory: InstanceFactory)
where
    InstanceFactory: TestInstanceFactory,
    InstanceFactory::Instance: InstanceForGetters,
    <<InstanceFactory::Instance as Instance>::Runtime as Runtime>::Memory:
        RuntimeMemory<InstanceFactory::Instance>,
{
    let instance = factory.load_test_module("getters");

    let mut getters = Getters::new(instance);

    assert_eq!(
        getters
            .get_true()
            .expect("Failed to run guest's `get-true` function"),
        true
    );
    assert_eq!(
        getters
            .get_false()
            .expect("Failed to run guest's `get-false` function"),
        false
    );
    assert_eq!(
        getters
            .get_s8()
            .expect("Failed to run guest's `get-s8` function"),
        -125
    );
    assert_eq!(
        getters
            .get_u8()
            .expect("Failed to run guest's `get-u8` function"),
        200
    );
    assert_eq!(
        getters
            .get_s16()
            .expect("Failed to run guest's `get-s16` function"),
        -410
    );
    assert_eq!(
        getters
            .get_u16()
            .expect("Failed to run guest's `get-u16` function"),
        60_000
    );
    assert_eq!(
        getters
            .get_s32()
            .expect("Failed to run guest's `get-s32` function"),
        -100_000
    );
    assert_eq!(
        getters
            .get_u32()
            .expect("Failed to run guest's `get-u32` function"),
        3_000_111
    );
    assert_eq!(
        getters
            .get_s64()
            .expect("Failed to run guest's `get-s64` function"),
        -5_000_000
    );
    assert_eq!(
        getters
            .get_u64()
            .expect("Failed to run guest's `get-u64` function"),
        10_000_000_000
    );
    assert_eq!(
        getters
            .get_float32()
            .expect("Failed to run guest's `get-f32` function"),
        -0.125
    );
    assert_eq!(
        getters
            .get_float64()
            .expect("Failed to run guest's `get-f64` function"),
        128.25
    );
}

/// An interface to import functions with parameters.
#[wit_import(package = "witty-macros:test-modules")]
trait Setters {
    fn set_bool(value: bool);
    fn set_s8(value: i8);
    fn set_u8(value: u8);
    fn set_s16(value: i16);
    fn set_u16(value: u16);
    fn set_s32(value: i32);
    fn set_u32(value: u32);
    fn set_s64(value: i64);
    fn set_u64(value: u64);
    fn set_float32(value: f32);
    fn set_float64(value: f64);
}

/// Test importing functions with parameters.
#[test_case(MockInstanceFactory::default(); "with a mock instance")]
fn setters<InstanceFactory>(mut factory: InstanceFactory)
where
    InstanceFactory: TestInstanceFactory,
    InstanceFactory::Instance: InstanceForSetters,
    <<InstanceFactory::Instance as Instance>::Runtime as Runtime>::Memory:
        RuntimeMemory<InstanceFactory::Instance>,
{
    let instance = factory.load_test_module("setters");

    let mut setters = Setters::new(instance);

    setters
        .set_bool(false)
        .expect("Failed to run guest's `set-bool` function");
    setters
        .set_s8(-100)
        .expect("Failed to run guest's `set-s8` function");
    setters
        .set_u8(201)
        .expect("Failed to run guest's `set-u8` function");
    setters
        .set_s16(-20_000)
        .expect("Failed to run guest's `set-s16` function");
    setters
        .set_u16(50_000)
        .expect("Failed to run guest's `set-u16` function");
    setters
        .set_s32(-2_000_000)
        .expect("Failed to run guest's `set-s32` function");
    setters
        .set_u32(4_000_000)
        .expect("Failed to run guest's `set-u32` function");
    setters
        .set_s64(-25_000_000_000)
        .expect("Failed to run guest's `set-s64` function");
    setters
        .set_u64(7_000_000_000)
        .expect("Failed to run guest's `set-u64` function");
    setters
        .set_float32(10.5)
        .expect("Failed to run guest's `set-f32` function");
    setters
        .set_float64(-0.000_08)
        .expect("Failed to run guest's `set-f64` function");
}
