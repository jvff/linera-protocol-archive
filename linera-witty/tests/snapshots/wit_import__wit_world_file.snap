---
source: linera-witty/tests/wit_import.rs
expression: "WitWorldWriter::new(\"witty-macros:test-modules\",\n                            \"test-world\").export::<SimpleFunction<MockInstance<()>>>().export::<Getters<MockInstance<()>>>().export::<Setters<MockInstance<()>>>().export::<Operations<MockInstance<()>>>().generate_file_contents().collect::<String>()"
---
package witty-macros:test-modules;

world test-world {
    export simple-function;
    export getters;
    export setters;
    export operations;
}
